src/libostree/ostree-repo-static-delta-processing.c

```diff
@@ -423,14 +456,48 @@ dispatch_bspatch (OstreeRepo *repo, StaticDeltaExecutionState *state, GCancellab
       struct bspatch_stream stream;
       stream.read = bspatch_read;
       stream.opaque = &opaque;
-      if (bspatch ((const guint8 *)g_mapped_file_get_contents (input_mfile),
-                   g_mapped_file_get_length (input_mfile), buf, state->content_size, &stream)
-          < 0)
+
+      int bspatch_result
+          = bspatch ((const guint8 *)g_mapped_file_get_contents (input_mfile),
+                     g_mapped_file_get_length (input_mfile), buf, state->content_size, &stream);
+
+      /* Unmap before checking result - only if we actually created a mapping */
+      if (state->content_size > 0 && munmap (buf, state->content_size) < 0)
+        return glnx_throw_errno_prefix (error, "munmap");
+
+      if (bspatch_result < 0)
         return glnx_throw (error, "bsdiff patch failed");
 
-      if (!_ostree_repo_bare_content_write (repo, &state->content_out, buf, state->content_size,
-                                            cancellable, error))
-        return FALSE;
+      /* Now read from tmpfile and write to repository (if content_size > 0) */
+      if (state->content_size > 0)
+        {
+          if (lseek (tmpf.fd, 0, SEEK_SET) < 0)
+            return glnx_throw_errno_prefix (error, "lseek");
+
+          /* Read and write in chunks to avoid loading entire file into memory */
+          guint64 bytes_remaining = state->content_size;
+          while (bytes_remaining > 0)
+            {
+              guchar chunk_buf[8192];
```

Comment:
Have you considered bigger block sizes like experimented with in `src/libostree/ostree-repo-commit.c:1232`?